### ISSUE TITLE
False positive with `callback-return` rule

### ISSUE THREAD

**Issue by sindresorhus** [eslint/eslint#3420]
ESLint 1.1.0. Concerning the [`callback-return`](http://eslint.org/docs/rules/callback-return) rule.

## Config

```
'callback-return': [2, ['cb', 'callback']]
```

## Problem

```js
function foo(bool, bool2, cb) {
	function bar() {
		if (bool2) {
			console.log('foo');
			cb();
		}
	}

	if (bool) {
		bar();
	} else {
		cb();
	}
}
```

I have many instances of this and similar code where I have a conditional whether to do something more or callback right away.

In the above example, both of the `cb` calls violates the `callback-return` rule. It should not however, as there is no code running after the `cb()` calls. So it's totally safe.

-

Some real examples:

- https://github.com/sindresorhus/gulp-size/blob/40c6c9af54046b7b45cf5b2e3df128da8236306c/test.js#L21
- https://github.com/sindresorhus/gulp-size/blob/40c6c9af54046b7b45cf5b2e3df128da8236306c/test.js#L82
- https://github.com/sindresorhus/gulp-size/blob/40c6c9af54046b7b45cf5b2e3df128da8236306c/test.js#L104

**Comment by eslintbot** [comment]
Thanks for the issue! We get a lot of issues, so this message is automatically posted to each one to help you check that you've included all of the information we need to help you.

Reporting a bug? Please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
1. The source code that caused the problem
1. The configuration you're using (for the rule or your entire config file)
1. The actual ESLint output complete with line numbers

Requesting a new rule? Please be sure to include:

1. The use case for the rule - what is it trying to prevent or flag?
1. Whether the rule is trying to prevent an error or is purely stylistic
1. Why you believe this rule is generic enough to be included

Requesting a feature? Please be sure to include:

1. The problem you want to solve (don't mention the solution)
1. Your take on the correct solution to problem

Including this information in your issue helps us to triage it and get you a response as quickly as possible.

Thanks!

**Comment by gyandeeps** [comment]
Can you please share the exact output of `eslint` for the above code?

**Comment by sindresorhus** [comment]
```
   5:4  error  Expected return with your callback function  callback-return
  12:3  error  Expected return with your callback function  callback-return
```

**Comment by gyandeeps** [comment]
Based on the [docs](http://eslint.org/docs/rules/callback-return#ifelse-statements), it has already discussed the case of if-else where the rule will warn. Why not just turn off the rule?

**Comment by xjamundx** [comment]
Yes, this is a known issue. I get where @sindresorhus is coming from: it's not great. This was primarily done to simplify the rule. Things were about to get very confusing and I 'd previously stumbled trying to create a rule that accurately accounted for those many edge cases. There even a bigger missing one which is the `process.nextTick(next)` situation, which is completely ignored currently. Those are pretty much all documented here:
http://eslint.org/docs/rules/callback-return#gotchas

With more time we could probably carefully account for each, but personally I've found it easier to just go along with the stricter rule and just `return` everywhere :)

**Here's my thought**: if someone is interested in taking this on these enhancements as a plugin, I'd be happy to help them with it. Once it's been tested pretty thoroughly in that context we can think about merging those back in to the main rule, but I think having a simplified rule which offers known benefits is a good start.

Thoughts from anyone else?

**Comment by sindresorhus** [comment]
> Why not just turn off the rule?

The rule has saved me multiple times, so I would like to use it.

> With more time we could probably carefully account for each, but personally I've found it easier to just go along with the stricter rule and just return everywhere :)

I'd rather just turn off the rule than add moot code just to please ESLint.

**Comment by nzakas** [comment]
In order to do this, we'd need path analysis, which we don't have right now (though it's being discussed #3373). So in the near-term, your only options are to turn off the rule, manually disable around things you know are okay, or add a `return`. When/if path analysis arrives, we can revisit this.

**Comment by ilyavolodin** [comment]
@mysticatea I can't remember if `callback-return` has been upgraded to use code path analysis, if it has, could you close this issue as resolved? If not, we should probably mark this as accepted and update the rule.

**Comment by mysticatea** [comment]
I had not touched this rule before.
I can investigate this in future.

**Comment by kaicataldo** [comment]
I took a stab at this - I tried to do it with code path analysis but wasn't able to get it working - maybe I was doing it wrong?

Thoughts on this attempt? Can add more tests and things, but wanted to get some feedback first: https://github.com/eslint/eslint/commit/d526db5d67fb3c8cf9ecd579378fa0bebfd6abcd

**Comment by nzakas** [comment]
@kaicataldo unfortunately, I have no idea how to do this. That's @mysticatea's area of expertise.

**Comment by mysticatea** [comment]
Oh, I'm sorry for late.

I'm not sure whether the approach is good or not as I'm not familiar with this rule.
If you use code path analysis, it would be similar to [constructor-super]. `constructor-super` is checking `super()` appears exact 1 time on each path if the class has a base class. That rule is complex for handling loops, though. Also, to satisfy this request, we should check inner functions in total. It may be difficult.

[constructor-super]: https://github.com/eslint/eslint/blob/1cc4356c50656a22106ba404b8ecea4d411aa1f5/lib/rules/constructor-super.js

**Comment by nzakas** [comment]
This has been open for over a year and several team members have indicated that this is quite difficult, so closing.

**Comment by xjamundx** [comment]
I'd really love to figure this one out as I think it can add a ton of value. Just leave it closed and if I can figure out how to do it with the code path stuff I'll shout.
